applications/FluidDynamicsApplication/custom_utilities/two_fluid_navier_stokes_data.h

```diff
@@ -0,0 +1,398 @@
+	const BoundedMatrix<double, TNumNodes, TDim>& rDN_DXenr)
+{
+	FluidElementData<TDim, TNumNodes, true>::UpdateGeometryValues(IntegrationPointIndex, NewWeight, rN, rDN_DX);
+	ElementSize = ElementSizeCalculator<TDim, TNumNodes>::GradientsElementSize(rDN_DX);
+	noalias(this->Nenr) = rNenr;
+	noalias(this->DN_DXenr) = rDN_DXenr;
+}
+
+static int Check(const Element& rElement, const ProcessInfo& rProcessInfo)
+{
+    const Geometry< Node<3> >& r_geometry = rElement.GetGeometry();
+    
+    KRATOS_CHECK_VARIABLE_KEY(VELOCITY);
+	KRATOS_CHECK_VARIABLE_KEY(DISTANCE);
+    KRATOS_CHECK_VARIABLE_KEY(MESH_VELOCITY);
+    KRATOS_CHECK_VARIABLE_KEY(BODY_FORCE);
+    KRATOS_CHECK_VARIABLE_KEY(PRESSURE);
+
+    for (unsigned int i = 0; i < TNumNodes; i++)
+    {
+        KRATOS_CHECK_VARIABLE_IN_NODAL_DATA(VELOCITY,r_geometry[i]);
+		KRATOS_CHECK_VARIABLE_IN_NODAL_DATA(DISTANCE, r_geometry[i]);
+        KRATOS_CHECK_VARIABLE_IN_NODAL_DATA(MESH_VELOCITY,r_geometry[i]);
+        KRATOS_CHECK_VARIABLE_IN_NODAL_DATA(BODY_FORCE,r_geometry[i]);
+        KRATOS_CHECK_VARIABLE_IN_NODAL_DATA(PRESSURE,r_geometry[i]);
+    }
+
+    KRATOS_CHECK_VARIABLE_KEY(DENSITY);
+    KRATOS_CHECK_VARIABLE_KEY(DYNAMIC_VISCOSITY);
+    KRATOS_CHECK_VARIABLE_KEY(DELTA_TIME);
+    KRATOS_CHECK_VARIABLE_KEY(DYNAMIC_TAU);
+    KRATOS_CHECK_VARIABLE_KEY(BDF_COEFFICIENTS);
+
+    return 0;
+}
+
+bool IsCut() {
+	return (NumPositiveNodes > 0) && (NumNegativeNodes > 0);
+}
+
+bool IsAir() {
+	return (NumPositiveNodes == TNumNodes);
+}
+
+void CalculateAirMaterialResponse() {
+	unsigned int strain_size;
+
+	if (TDim == 2)
+		strain_size = 3;
+	else if (TDim == 3)
+		strain_size = 6;
+
+	if(this->C.size1() != strain_size)
+		this->C.resize(strain_size,strain_size,false);
+	if(this->ShearStress.size() != strain_size)
+		this->ShearStress.resize(strain_size,false);
+
+    ComputeStrain(); 
+
+    CalculateEffectiveViscosityAtGaussPoint();
+
+	const double mu = this->EffectiveViscosity;
+	const double c1 = 2.0*mu;
+	const double c2 = mu;
+
+	this->C.clear();
+    Matrix& c_mat = this->C;
+    Vector& stress = this->ShearStress;
+    Vector& strain = this->StrainRate;
+
+	if (TDim == 2) 
+	{
+        constexpr double two_thirds = 2./3.;
+        constexpr double four_thirds = 4./3.;
+
+        c_mat(0,0) = mu * four_thirds;
+        c_mat(0,1) = -mu * two_thirds;
+        c_mat(0,2) = 0.0;
+        c_mat(1,0) = -mu * two_thirds;
+        c_mat(1,1) = mu * four_thirds;
+        c_mat(1,2) = 0.0;
+        c_mat(2,0) = 0.0;
+        c_mat(2,1) = 0.0;
+        c_mat(2,2) = mu;
+
+		c_mat(0, 0) = 2.0*mu;
+		c_mat(1, 1) = 2.0*mu;
+		c_mat(2, 2) = mu;
+
+        const double trace = strain[0] + strain[1];
+        const double volumetric_part = trace/2.0; // Note: this should be small for an incompressible fluid (it is basically the incompressibility error)
+
+		stress[0] = c1 * strain[0] - volumetric_part; 
+		stress[1] = c1 * strain[1] - volumetric_part;
+		stress[2] = c2 * strain[2];
+	}
```

Comment:
There are methods to calculate C in `fluid_element_utilities`, I would try to use those if possible (to avoid code duplication). If they are not usable, please let me know why and we can try to improve them.

Reply:
done !